src/test/nodesessionmetadata_test.cpp

```diff
@@ -0,0 +1,215 @@
+    NodeSessionMetadata meta;
+    auto demultiplexedMetas = meta.generateSubsessions("request", 2);
+    ASSERT_EQ(demultiplexedMetas.size(), 2);
+    EXPECT_EQ(demultiplexedMetas[0].getSessionKey(), "request_0");
+    EXPECT_EQ(demultiplexedMetas[1].getSessionKey(), "request_1");
+}
+TEST_F(NodeSessionMetadataTest, GenerateTwoLevelsOfSubsession) {
+    const uint firstLevelDemultiplexSize = 3;
+    const uint secondLevelDemultiplexSize = 2;
+    NodeSessionMetadata meta;
+    auto demultiplexedMetas = meta.generateSubsessions("request", firstLevelDemultiplexSize);
+    ASSERT_EQ(demultiplexedMetas.size(), firstLevelDemultiplexSize);
+    std::vector<NodeSessionMetadata> secondLevelMetas(firstLevelDemultiplexSize * secondLevelDemultiplexSize);
+    for (size_t demMetaId = 0; demMetaId != demultiplexedMetas.size(); ++demMetaId) {
+        auto newLevelMetas = demultiplexedMetas[demMetaId].generateSubsessions("2ndDemultiplexer", secondLevelDemultiplexSize);
+        std::move(newLevelMetas.begin(), newLevelMetas.end(), secondLevelMetas.begin() + demMetaId * secondLevelDemultiplexSize);
+    }
+    for (size_t demMetaId = 0; demMetaId != demultiplexedMetas.size(); ++demMetaId) {
+        EXPECT_EQ(demultiplexedMetas[demMetaId].getSessionKey(), std::string("request_") + std::to_string(demMetaId));
+    }
+    for (size_t demMetaId = 0; demMetaId != firstLevelDemultiplexSize; ++demMetaId) {
+        for (size_t demMetaLev2Id = 0; demMetaLev2Id != secondLevelDemultiplexSize; ++demMetaLev2Id) {
+            auto hash = secondLevelMetas[demMetaLev2Id + demMetaId * secondLevelDemultiplexSize].getSessionKey();
+            std::string searchFor = std::string("request_") + std::to_string(demMetaId);
+            EXPECT_NE(hash.find(searchFor), std::string::npos) << hash << " searching for:" << searchFor;
+            searchFor = std::string("2ndDemultiplexer_") + std::to_string(demMetaLev2Id);
+            EXPECT_NE(hash.find(searchFor), std::string::npos) << hash << " searching for:" << searchFor;
+        }
+    }
+}
+TEST_F(NodeSessionMetadataTest, GenerateThreeLevelsOfSubsession) {
+    const uint firstLevelDemultiplexSize = 3;
+    const uint secondLevelDemultiplexSize = 2;
+    const uint thirdLevelDemultiplexSize = 4;
+    NodeSessionMetadata meta;
+    auto demultiplexedMetaLev3 = meta
+                                     .generateSubsessions("request", firstLevelDemultiplexSize)[2]
+                                     .generateSubsessions("extract1st", secondLevelDemultiplexSize)[0]
+                                     .generateSubsessions("extract2nd", thirdLevelDemultiplexSize)[2];
+    auto hash = demultiplexedMetaLev3.getSessionKey();
+    std::string searchFor = "request_2";
+    EXPECT_NE(hash.find(searchFor), std::string::npos) << hash << " searching for:" << searchFor;
+    searchFor = "extract1st_0";
+    EXPECT_NE(hash.find(searchFor), std::string::npos) << hash << " searching for:" << searchFor;
```

Comment:
Use `EXPECT_THAT(returnedString, HasSubstr("word"));` instead of `EXPECT_NE` & `std::string::npos`

Reply:
This way we can get rid of `searchFor` variable in each unit test therefore reduce code by a lot.